File: apps/playbooks-ai/rules-engine/math-ai/benchmark.ts

import { MathEngine } from './engine';

const engine = new MathEngine();
const expression = "1 + 2(30+19)/3^2 + 1000 + 1 + 2(30+19)/3^2 + 1000 - 1 + 2(30+19)/3^2 + 1000 / (1 + 2(30+19)/3^2 + 1000) * 21  - 1 + 2(30+19)/3^2 + 1000 + 10002";
const ITERATIONS = 100_000;

console.log(`\x1b[1m🚀 TS BENCHMARK: ${ITERATIONS.toLocaleString()} iterations\x1b[0m`);

// Pre-parse to measure pure execution speed
const model = engine.parseExpression(expression);

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
   // Use a fresh engine instance or clear results to ensure fair work
   const runner = new MathEngine({}, true);
   runner.run(model);
}
const end = performance.now();

const totalMs = end - start;
const avgMs = totalMs / ITERATIONS;

console.log(`\n  Total Time: \x1b[36m${totalMs.toFixed(2)}ms\x1b[0m`);
console.log(`  Avg/Run:    \x1b[36m${(avgMs * 1000).toFixed(4)}μs\x1b[0m`);
console.log(`  Ops/Sec:    \x1b[32m${Math.round(ITERATIONS / (totalMs / 1000)).toLocaleString()}\x1b[0m`);

File: apps/playbooks-ai/rules-engine/math-ai-rust/src/bin/benchmark.rs
use math_ai_rust::{MathEngine};
use std::time::Instant;

fn main() {
    let expression = "1 + 2(30+19)/3^2 + 1000 + 1 + 2(30+19)/3^2 + 1000 - 1 + 2(30+19)/3^2 + 1000 / (1 + 2(30+19)/3^2 + 1000) * 21  - 1 + 2(30+19)/3^2 + 1000 + 10002";
    let iterations = 100_000;

    println!("\x1b[1m🦀 RUST BENCHMARK: {} iterations\x1b[0m", iterations);

    let engine = MathEngine::new(true);
    let wrapper = engine.parse_expression(expression).unwrap();

    let mut runner = MathEngine::new(true);
    let start = Instant::now();
    for _ in 0..iterations {
        runner.clear_variables();
        let _ = runner.run(&wrapper);
    }
    let duration = start.elapsed();

    let total_ms = duration.as_secs_f64() * 1000.0;
    let avg_us = (total_ms * 1000.0) / iterations as f64;
    let ops_per_sec = iterations as f64 / duration.as_secs_f64();

    println!("\n  Total Time: \x1b[36m{:.2}ms\x1b[0m", total_ms);
    println!("\n  Avg/Run:    \x1b[36m{:.4}μs\x1b[0m", avg_us);
    println!("  Ops/Sec:    \x1b[32m{}\x1b[0m", (ops_per_sec as u64));
}
